### DISCUSSION TITLE
How to install and configure on a Raspberry Pi5

### DISCUSSION THREAD

**Discussion by Arelius-D** [mvp/uhubctl#563] [General]
_Please do laugh at me if you have to.._ my question may sound almost stupid maybe but I really don't know how I could install uhubctl and here asking for help if possible. Is there some documentation around that here somewhere? 

_I have running Raspberry Pi OS Lite 64bit Debian version: 12 (bookworm) on a Raspberry Pi 5._

**Comment by mvp** [comment]
uhubctl is included in standard OS packages and can installed with:

```
sudo apt install uhubctl 
```

However, currently this would install version 2.5.0, which works fine for all platforms, but doesn't yet have full support for RPi5 baked in.

Until I release new version 2.6.0 which will support RPi5, you can use install from latest source code as described in readme:

https://github.com/mvp/uhubctl?tab=readme-ov-file#compiling:


```
sudo apt-get install libusb-1.0-0-dev
git clone https://github.com/mvp/uhubctl
cd uhubctl
make
sudo make install
```

Also, please read instructions on how to control RPi5 ports in readme - it is somewhat weird

**Reply by Arelius-D** [reply]
Yeah read that part cause it was on the main page and got scared enough 😅
May I ask, is it possible to put the commands in a file and run it as an executable which would possibly run itself automatically upon each boot/startup?

I have been struggling with my problem for weeks and gone down so many rabbit holes with no success..

Latest one beeing [this](https://unix.stackexchange.com/questions/775387/compile-a-script-add-it-to-system-so-it-runs-before-or-during-system-boot)

**Reply by mvp** [reply]
If you need to fully reset all USB devices at boot time on RPi5, install uhubctl from master branch as described above, then add following commands to some startup script:

```
uhubctl -l 1 -a 0
uhubctl -l 3 -a 0 
uhubctl -l 1 -a 1
uhubctl -l 3 -a 1
```
They need to be executed as root, with sudo or with USB permissions fixed.
Also, in startup script you may need to use full path to uhubctl executable if your PATH is not properly configured.

**Reply by Arelius-D** [reply]
ATM my problem is with my only device attached to the same USB3 port so I would prefer to only run the commands targeted for that port but as far I can tell from what I read that wouldn’t be possible(?)

```
sudo lsusb
Bus 004 Device 001: ID 1d6b:0003 Linux Foundation 3.0 root hub
Bus 003 Device 001: ID 1d6b:0002 Linux Foundation 2.0 root hub
Bus 002 Device 003: ID 0bda:8156 Realtek Semiconductor Corp. USB 10/100/1G/2.5G LAN
Bus 002 Device 001: ID 1d6b:0003 Linux Foundation 3.0 root hub
Bus 001 Device 001: ID 1d6b:0002 Linux Foundation 2.0 root hub

sudo lsusb -t
/:  Bus 04.Port 1: Dev 1, Class=root_hub, Driver=xhci-hcd/1p, 5000M
/:  Bus 03.Port 1: Dev 1, Class=root_hub, Driver=xhci-hcd/2p, 480M
/:  Bus 02.Port 1: Dev 1, Class=root_hub, Driver=xhci-hcd/1p, 5000M
    |__ Port 1: Dev 3, If 0, Class=Vendor Specific Class, Driver=r8152, 5000M
/:  Bus 01.Port 1: Dev 1, Class=root_hub, Driver=xhci-hcd/2p, 480M

ls -l /sys/bus/usb/devices/
total 0
lrwxrwxrwx 1 root root 0 Apr 28 14:24 1-0:1.0 -> ../../../devices/platform/axi/1000120000.pcie/1f00200000.usb/xhci-hcd.0/usb1/1-0:1.0
lrwxrwxrwx 1 root root 0 Apr 28 14:24 2-0:1.0 -> ../../../devices/platform/axi/1000120000.pcie/1f00200000.usb/xhci-hcd.0/usb2/2-0:1.0
lrwxrwxrwx 1 root root 0 Apr 28 14:24 2-1 -> ../../../devices/platform/axi/1000120000.pcie/1f00200000.usb/xhci-hcd.0/usb2/2-1
lrwxrwxrwx 1 root root 0 Apr 28 14:24 2-1:1.0 -> ../../../devices/platform/axi/1000120000.pcie/1f00200000.usb/xhci-hcd.0/usb2/2-1/2-1:1.0
lrwxrwxrwx 1 root root 0 Apr 28 14:24 3-0:1.0 -> ../../../devices/platform/axi/1000120000.pcie/1f00300000.usb/xhci-hcd.1/usb3/3-0:1.0
lrwxrwxrwx 1 root root 0 Apr 28 14:24 4-0:1.0 -> ../../../devices/platform/axi/1000120000.pcie/1f00300000.usb/xhci-hcd.1/usb4/4-0:1.0
lrwxrwxrwx 1 root root 0 Apr 28 14:24 usb1 -> ../../../devices/platform/axi/1000120000.pcie/1f00200000.usb/xhci-hcd.0/usb1
lrwxrwxrwx 1 root root 0 Apr 28 14:24 usb2 -> ../../../devices/platform/axi/1000120000.pcie/1f00200000.usb/xhci-hcd.0/usb2
lrwxrwxrwx 1 root root 0 Apr 28 14:24 usb3 -> ../../../devices/platform/axi/1000120000.pcie/1f00300000.usb/xhci-hcd.1/usb3
lrwxrwxrwx 1 root root 0 Apr 28 14:24 usb4 -> ../../../devices/platform/axi/1000120000.pcie/1f00300000.usb/xhci-hcd.1/usb4
```

**Reply by Arelius-D** [reply]
Ok so here is what I did,

I ran
```ip a > status_before_uhubctl.txt```

then did
```sudo uhubctl -l 1 -a 0 > output_of_uhubctl.txt```

and then
```ip a > status_after_uhubctl.txt```


(I did this for all for commands, check the outputs and discovered that in my case ```sudo uhubctl -l 1 -a 0``` did the trick!)

**Reply by mvp** [reply]
If you want to truly toggle power for all connected USB devices, you have to execute these commands:
```
uhubctl -l 1 -a 0
uhubctl -l 3 -a 0 
uhubctl -l 1 -a 1
uhubctl -l 3 -a 1
```
Running it against only one bus say 1 may make your misbehaving device appear to work again by having it reconnect. However, that will not guarantee full power reset. Due to how RPi5 hardware design, all USB ports are ganged into one power source, and the only way to truly power it off is to power off all ports.

**Reply by Arelius-D** [reply]
So even if I use one USB device connected to the same port (_don't even remove it ever but just to clarify_) it's still best or rather recommended that all USB ports should be cycled throw cause of this particular design of this SBC? If so, will this make me "safe"?

```
[Unit]
Description=Power Cycle of specific USB port

[Service]
ExecStart=uhubctl -l 1 -a 0 ; uhubctl -l 3 -a 0 ; uhubctl -l 1 -a 1 ; uhubctl -l 3 -a 1
Type=oneshot
RemainAfterExit=yes

[Install]
WantedBy=multi-user.target
```

**Reply by mvp** [reply]
Yes, that is correct - all 4 onboard USB ports in RPi5 are ganged into one group and need to be reset together to have true power cycle.

If you need to target ports separately for power cycling, you can attach any hub from supported devices table to RPi5 and then you would be able to control power individually. In particular, good option is Uugear Mega4 - this 4 port USB3 hub supports per port power switching and was designed for RPi4, but also works with RPi5 by attaching on top of it.

**Reply by Arelius-D** [reply]
Thank you for the clarification, and if ok, I would like to share this [link](https://unix.stackexchange.com/questions/775387/compile-a-script-add-it-to-system-so-it-runs-before-or-during-system-boot) maybe it would be helpful for someone else who just like myself os out if his depth.

Without **uhubctl** I would have not solve my problem! So thank you again.

Also, cheers for the mention of the Uugear Mega4, I have to wait for a version which can be mounted underneath though since I'm using HAT for SSD for boot/System and some of my Containers and so on.

**Reply by mvp** [reply]
Glad you have your problem addressed. Note that Uugear doesn't really have to be attached underneath, you can treat it as standard USB hub and connect it using USB cable. However, when you connect it on top it looks nicer and it becomes one convenient package. Regarding possible improvements of Uugear for HAT and RPi5 specific compatibility, I recommend reaching out to @uugear, he is very helpful and responsive.

**Comment by Arelius-D** [comment]
> sudo apt-get install libusb-1.0-0-dev
> git clone https://github.com/mvp/uhubctl
> cd uhubctl
> make
> sudo make install

I just tried to install this and there seems to be something wrong with the second command

```
pi@Pi5:~ $ sudo git clone https://github.com/mvp/uhubctl
sudo: git: command not found
```

**Reply by Arelius-D** [reply]
Ok this I did figure out, had to do it like this

```
sudo apt-get update
sudo apt-get install git -y


git config --global user.name "my_git_username"
git config --global user.email "my_email@address"

sudo apt-get install libusb-1.0-0-dev
sudo git clone https://github.com/mvp/uhubctl
cd uhubctl
sudo make
sudo make install
```

**Reply by jaycrav3ns** [reply]
Like I just said on Stack, I could tell you were serious.  Good stuff!  Give me a couple hours to finish up here, I'll take a look at your major update. 👍
